sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java

```diff
@@ -439,4 +652,553 @@ public void process(ProcessContext c) {
+     * the collection is empty, to generate the (only, empty) output file.
+     */
+    public Write<DestinationT, UserT> withEmptyGlobalWindowDestination(
+        DestinationT emptyWindowDestination) {
+      return toBuilder().setEmptyWindowDestination(emptyWindowDestination).build();
+    }
+
+    /**
+     * Specifies a {@link Coder} for the destination type, if it can not be inferred from {@link
+     * #by}.
+     */
+    public Write<DestinationT, UserT> withDestinationCoder(Coder<DestinationT> destinationCoder) {
+      checkArgument(destinationCoder != null, "destinationCoder can not be null");
+      return toBuilder().setDestinationCoder(destinationCoder).build();
+    }
+
+    /**
+     * Specifies the side inputs that should be made available to all user code running as part of
+     * this transform.
+     */
+    public Write<DestinationT, UserT> withSideInputs(List<PCollectionView<?>> sideInputs) {
+      checkArgument(sideInputs != null, "sideInputs can not be null");
+      return toBuilder().setSideInputs(sideInputs).build();
+    }
+
+    /** Like {@link #withSideInputs(List)}. */
+    public Write<DestinationT, UserT> withSideInputs(PCollectionView<?>... sideInputs) {
+      return toBuilder().setSideInputs(Arrays.asList(sideInputs)).build();
+    }
+
+    /**
+     * Specifies to use a given fixed number of shards per window. 0 means runner-determined
+     * sharding. Specifying a non-zero value may hurt performance, because it will limit the
+     * parallelism of writing and will introduce an extra {@link GroupByKey} operation.
+     */
+    public Write<DestinationT, UserT> withNumShards(int numShards) {
+      checkArgument(numShards >= 0, "numShards must be non-negative, but was: %s", numShards);
+      if (numShards == 0) {
+        return withNumShards(null);
+      }
+      return withNumShards(ValueProvider.StaticValueProvider.of(numShards));
+    }
+
+    /**
+     * Like {@link #withNumShards(int)}. Specifying {@code null} means runner-determined sharding.
+     */
+    public Write<DestinationT, UserT> withNumShards(ValueProvider<Integer> numShards) {
+      return toBuilder().setNumShards(numShards).build();
+    }
+
+    /**
+     * Specifies a {@link PTransform} to use for computing the desired number of shards in each
+     * window.
+     */
+    public Write<DestinationT, UserT> withSharding(
+        PTransform<PCollection<UserT>, PCollectionView<Integer>> sharding) {
+      checkArgument(sharding != null, "sharding can not be null");
+      return toBuilder().setSharding(sharding).build();
+    }
+
+    /**
+     * Specifies to ignore windowing information in the input, and instead rewindow it to global
+     * window with the default trigger.
+     *
+     * @deprecated Avoid usage of this method: its effects are complex and it will be removed in
+     *     future versions of Beam. Right now it exists for compatibility with {@link WriteFiles}.
```

Comment:
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
There are specific semantic differences today (e.g. what is done with empty shards), so I'm not sure its use can be avoided. Maybe we'll be able to sort these semantic issues out in the future, but I don't think we are there yet.